cognite_toolkit/_cdf_tk/loaders/_resource_loaders/timeseries_loaders.py

```diff
@@ -452,16 +452,115 @@ def split_timeseries_ids(
+            elif ts.instance_id and ts.external_id is None:
+                current_instance_ids.add(ts.instance_id)
+            elif ts.external_id and ts.instance_id:
+                # Migrated time series with both external_id and instance_id
+                if ts.external_id in desired_timeseries_ids and ts.instance_id not in desired_instance_ids:
+                    current_timeseries_ids.add(ts.external_id)
+                elif ts.external_id not in desired_timeseries_ids and ts.instance_id in desired_instance_ids:
+                    current_instance_ids.add(ts.instance_id)
+                elif ts.external_id in desired_timeseries_ids and ts.instance_id in desired_instance_ids:
+                    current_timeseries_ids.add(ts.external_id)
+                    current_instance_ids.add(ts.instance_id)
+                else:
+                    # It is in neither of the desired sets, so it will be removed.
+                    # We use instanceId as a preference to avoid duplicates.
+                    current_instance_ids.add(ts.instance_id)
+
+        # Calculate what needs to be added and removed
+        ts_to_add = desired_timeseries_ids - current_timeseries_ids
+        ts_to_remove = current_timeseries_ids - desired_timeseries_ids
+        instance_to_add = desired_instance_ids - current_instance_ids
+        instance_to_remove = current_instance_ids - desired_instance_ids
+
+        # Clear the time series IDs from the main update to avoid conflicts
+        # The to_update object is used to update all other properties of the subscription.
+        to_update.time_series_ids = None
+        to_update.instance_ids = None
+
+        # Create update batches for changes
+        batches: list[DataPointSubscriptionUpdate] = []
+        all_removals = [("ts", id_) for id_ in ts_to_remove] + [("instance", id_) for id_ in instance_to_remove]
+        all_additions = [("ts", id_) for id_ in ts_to_add] + [("instance", id_) for id_ in instance_to_add]
+        for removals, additions in zip_longest(
+            chunker(all_removals, cls._TIMESERIES_ID_REQUEST_LIMIT),
+            chunker(all_additions, cls._TIMESERIES_ID_REQUEST_LIMIT),
+            fillvalue=None,
+        ):
+            update = DataPointSubscriptionUpdate(external_id=subscription.external_id)
+            ts_ids_to_remove, instance_ids_to_remove = cls._split_ts_instance_ids(removals or [])
+            if ts_ids_to_remove:
+                update.time_series_ids.remove(ts_ids_to_remove)
+            if instance_ids_to_remove:
+                update.instance_ids.remove(instance_ids_to_remove)
+
+            ts_ids_to_add, instance_ids_to_add = cls._split_ts_instance_ids(additions or [])
+            if ts_ids_to_add:
+                update.time_series_ids.add(ts_ids_to_add)
+            if instance_ids_to_add:
+                update.instance_ids.add(instance_ids_to_add)
+            batches.append(update)
+
+        return to_update, batches
```

Comment:
Looks like `to_update` contains zero added/removed time series, while the batches contains these updates. An optimization could be to also add/remove time series in `to_update`.

This function could return just a list of updates instead of a tuple of a write (which gets turned into an update) and a list of updates.

Anyway, I think that can be optimized later if needed 🤷‍♂️